Scan the exceptor configuration for unknown classes before applying

Minecrell opened the pull request:
This fixes the problem reported in #244 by adding a new deobfuscation step that will scan the generated classes and remove any invalid inner class references it finds.

cpw commented:
I don't think this is the right way to fix the problem.

Minecrell commented:
@cpw I agree, it should be probably fixed in Exceptor/MCInjector instead (or whereever this problem is coming from), but after talking with @AbrarSyed we figured this might be the best temporary solution for now.

cpw commented:
Well, I think we need to figure that out. Exceptor and mcinjector are both projects we can fix..

Minecrell commented:
@cpw I have changed the fix to scan the provided exceptor configuration for unknown classes instead, and remove them if there are any. While this is something that could be also fixed in Exceptor itself in my opinion it should be probably up to the user of Exceptor to make sure the configuration is valid.

So basically it will now go through the exceptor configuration, check if the classes and inner classes exist and remove them if not. Currently also the class entries itself are removed if the classes don't exist (so for example, client classes when decompiling only the server), however this could be also changed to simply skip them if needed. Shouldn't make a big difference.

LexManos commented:
THe exceptor config is valid you're just applying a merged config onto a unmerged jar.
This is a USER ERROR nothing to do with exceptor or FG or anything. This is on you.
However, if you want to filter them out, then do it properly in the https://github.com/ModCoderPack/MCInjector damn mcinjector project.

cpw commented:
Well, that explains the bug then Lex..

AbrarSyed commented:
So I mistakenly assumed that the EXC and SRG files could be shared between all the jars, including the server, client and merged jars. Apperantly I can only use the EXC on the merged and client jars, and this problem appears if I use it on the server jar.

The options to fix his problem are as follows, orderred by correctness.
1. ship multiple EXCs in the MCP zip and apply the correct EXC file as part of FG.
2. fix the issue in McInjector, so that it doesnt inject claasses into the jar that dont exist.
3. fix it in ForgeGradle like in this here PR.

Now, due to my time constraints, I opted to do number 3. Minecrell was nice enough to send a PR for it before I actually got a chance to do it myself. This is meant to be a stop-gap measure to make stuff work until a more complete fix can be made.
Number 1 required me to add special casing for the server jar, which would be hard to remove later, and generally be a pain. I would also have to talk to bspkrs and have the correct files added to the MCP zip FG downloads. Its much easier to add this as a single removable function as was done in this PR.
Number 2 would definitely be preferred, but I dont have the time to delve into the MC Injector code at this time. If someone could do that for us, thatd be great, but for the time being this current fix is good enough, and isnt all together ugly or anything either.

bloodmc commented:
@AbrarSyed any update on this? We would like to update Sponge to FG2. Thanks!

LexManos commented:
In before scala users have issues because scala is dumb and requires all ACCESSED classes to have a inner class attribute as well, so this won't work.

AbrarSyed commented:
well its still snapshots, if theydont like they can go back to the older stable.. We arnt focring everyone to magically update anymore :)
